vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1224 @@
+                writeit(ss, "stride_shift");
+            }
+            
+            writeit(b.max_nodes, "max_nodes");
+            writeit(b.learn_at_leaf, "learn_at_leaf");
+            writeit(b.oas, "oas");
+	        //writeit(b.leaf_example_multiplier, "leaf_example_multiplier")
+            writeitvar(b.nodes.size(), "nodes", n_nodes); 
+            writeit(b.max_num_labels, "max_number_of_labels");
+
+            if (read){
+                b.nodes.clear();
+                for (uint32_t i = 0; i < n_nodes; i++)
+                    b.nodes.push_back(node());
+            }
+            
+            //node  
+            for(uint32_t i = 0; i < n_nodes; i++){
+                save_load_node(b.nodes[i], model_file, read, text, msg);
+            }
+            //deal with examples:
+            writeitvar(b.examples.size(), "examples", n_examples);
+            if (read){
+                b.examples.clear();
+                for (uint32_t i = 0; i < n_examples; i++){
+                    example* new_ec = &calloc_or_throw<example>();
+                    b.examples.push_back(new_ec);
+                }
+            }
+            for (uint32_t i = 0; i < n_examples; i++)
+                save_load_example(b.examples[i], model_file, read, text, msg, b.oas);
+	    //cout<<"done loading...."<<endl;
+            
+            
+        }
+    }
+    //////////////////////////////End of Save & Load///////////////////////////////
+} //namespace
+
+base_learner* memory_tree_setup(options_i& options, vw& all)
+{
+    using namespace memory_tree_ns;
+	auto tree = scoped_calloc_or_throw<memory_tree>();
+    option_group_definition new_options("Memory Tree");
+
+    new_options.add(make_option("memory_tree", tree->max_nodes).keep().help("Make a memory tree with at most <n> nodes"))
+        .add(make_option("max_number_of_labels",tree->max_num_labels).default_value(10).help("max number of unique label"))
+        .add(make_option("leaf_example_multiplier",tree->leaf_example_multiplier).default_value(1).help("multiplier on examples per leaf (default = log nodes)"))
+        .add(make_option("Alpha", tree->alpha).default_value(0.1f).help("Alpha"))
+        .add(make_option("dream_repeats", tree->dream_repeats).default_value(1).help("number of dream operations per example (default = 1)"))
+        .add(make_option("top_K",tree->top_K).default_value(1).help("top K prediction error (default 1)"))
+        .add(make_option("learn_at_leaf",tree->learn_at_leaf).default_value(0).help("whether or not learn at leaf (defualt = True)"))
+        .add(make_option("oas", tree->oas).default_value(0).help("use oas at the leaf"))
+        .add(make_option("dream_at_update", tree->dream_at_update).default_value(0).help("turn on dream operations at reward based update as well"))
+        .add(make_option("online", tree->online).default_value(0).help("turn on dream operations at reward based update as well"));
+    options.add_and_parse(new_options);
+
```

Comment:
The issue with test 1 is here---it's inserting itself into every reduction rather than just the memory_tree ones.  You should test for memory_tree and if it's not present return nullptr.

Reply:
Along with above, this fixes the off-by-epsilon issue in Windows unit tests.

```suggestion
if (!tree->max_nodes)
{
  return nullptr;
}
```